rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+  //   1. Kernel with the given name is not found.
+  //   2. Attributes in op_kernel_construction don't match type constraints
+  //      for any of the kernels with this name.
+  //      Note that we consider a constraint to be "not matched" if attribute
+  //      it applies to is not in op_kernel_construction.
+  std::unique_ptr<TFRTOpKernel> CreateKernel(
+      StringPiece kernel_class_name,
+      TFRTOpKernelConstruction* op_kernel_construction) const;
+
+ private:
+  llvm::StringMap<std::vector<TFRTOpKernelReg>> factories_;
+};
+
+extern llvm::ManagedStatic<TFRTOpKernelFactories>
+    tfrt_forwarding_kernel_factories;
+```
+
+Similar to current TensorFlow kernel registartion, we will introduce a
+registration macro that adds a kernel to `TFRTOpKernelFactories`.
+
+```cpp
+#define REGISTER_KERNEL_FALLBACK_KERNEL(name, ...) \
+  REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ_HELPER(__COUNTER__, name, __VA_ARGS__)
+
+#define REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ_HELPER(ctr, name, ...) \
+  REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ(ctr, name, __VA_ARGS__)
+
+#define REGISTER_KERNEL_FALLBACK_KERNEL_UNIQ(ctr, name, ...)             \
+  static bool global_tfrt_forwarding_kernel_##ctr##_registered_ = []() { \
+    ::tensorflow::tfrt_forwarding_kernel_factories->RegisterFactory(     \
+        name, TFRTOpKernelReg([](TFRTOpKernelConstruction* construction) \
+                                  -> std::unique_ptr<TFRTOpKernel> {     \
+          return std::make_unique<__VA_ARGS__>(construction);            \
+        }));                                                             \
+    return true;                                                         \
+  }();
+```
+
+## Op registration
+
+To support type specification, we will also provide a minimal Op registry and
+corresponding macro `REGISTER_KERNEL_FALLBACK_OP`.
+
+## Kernel implementation
+
+TensorFlow kernels inherit from the
+[OpKernel](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=82?q=opkernel)
+class and depend on two key classes:
+[OpKernelConstruction](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=256?q=opkernel)
+and
+[OpKernelContext](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.h;l=584?q=opkernel).
+We want to provide custom implementations of these two classes in terms of data
+we get from TFRT (for e.g. inputs, attributes).
+
+There are two main approaches to customize class implementations:
+
+*   Use inheritance and define common interfaces.
+*   Use templates.
+
+We ran multiple benchmarks to get an idea of the trade offs between inheritance
+and templating approaches. Key findings are summarized below:
+
+*   Time difference negligible for full model benchmarks.
+*   A simple scalar op benchmark with Kernel Fallback (runs scalar
+    multiplication, division, addition) was only 0.3% slower on mobile with
```

Comment:
worth stating the mobile architecture you run the benchmark on.

Reply:
Benchmark setup:

* Pixel 3
* ABI: arm64-v8a
* SDK version: 29

I added this info to the doc as well.